README.md

```diff
@@ -75,6 +75,12 @@ The recommended layout for your application is:
 
 All of these files should be checked into your application.
 
+The following diagram explains the above concept. As you can see in the diagram, `package-two`, which is a production dependency, and `package-three`, which is a dev dependency, depend on `package-eight`. But `package-eight` will only be listed in `requirements.txt`.
+
+![A package dependency diagram is shown. The dependency graph is as follows: 1. The top level application has two production dependencies, package-one and package-two, and two dev dependencies, package-three and package-four. 2. package-one in turn depends on package-five and package-six. 3. package-two depends on package-seven and package-eight. 4. package-three also depends on package-eight. 5. package-four only depends on package-nine.](https://docs.google.com/drawings/d/e/2PACX-1vTezwHVxvTdsL0Ta5ruvXFYBYw3XYlvcENlSOUe7wehLwb8yz-q4QKc5TZAm2FtYm0e92NrlgnSJLDL/pub?w=960&h=720)
+
+_The above diagram can be edited on [Google Drawings](https://docs.google.com/drawings/d/1FewA-gmh7vjuatHQlM_u7N56riyLfoH12huVOU0v1MA/edit)._
```

Comment:
instead of a google drawing please check in a file into this repository so anybody can edit / PR it.  An SVG would be preferred since it is text

I'm not sure the icons help the diagram either.  It might also be helpful to show where `==` lines come from.  I'd probably move the `-minimal` files out of the box in that case

Otherwise, looks great -- thanks for putting this together!

Reply:
(cc @adeng322)

> instead of a google drawing please check in a file into this repository so anybody can edit / PR it. An SVG would be preferred since it is text

We’ve updated the permissions on the Google Drawings to **Anyone who has the link can access. No sign-in required.** Let us know if that works instead.

As for Google Drawing vs SVG…
* The export link above remains the same even if the diagram is edited. Just edit the diagram and refresh the GitHub page. The exported image however gets cached by the browser though.
* Google Drawings has a WYSIWYG editor, while an [SVG exported from Google Drawings](https://i.fluffy.cc/6FtFQ2LVKfj4s7fktx2h0jV9KjH7XK8Z.html) isn’t really readable/editable. Also, creating a readable SVG by hand is not easy.

> I'm not sure the icons help the diagram either.

True, the icons are for aesthetics only, but they make the diagram look good!

> It might also be helpful to show where `==` lines come from. I'd probably move the `-minimal` files out of the box in that case

We’ve updated the diagram to reflect this. Please have a look.

Thanks!

Reply:
If you want a middle ground: export the google drawing, check that in, and use markdown something like:

```markdown
<!-- exported from google drawing at ... -->
![](./img/diagram.svg)
```

Mostly want the repository to not have a dependency on google drive which is likely to become inaccessible (and can't be edited by outsiders)

As for the diagram itself:
- `pip freeze` should be `upgrade-requirements`.
- where did the icons come from? are they compatible with the license of this repository?
- `requirements-dev.txt` should be in the top left corner like the rest of the boxes
- the spacing for the icons in the `requirements-dev.txt` box are off

(I'd just fix these things, but I can't because it's a google drawing -- this is why I don't want it to be a google drawing!!!)

Are there other alternatives? Something like [plantuml](http://plantuml.com/)? (ugly, but there's probably better alternatives / themese)?  I didn't realize hand-SVG was that difficult :S isn't it basically just html?

Reply:
> where did the icons come from? are they compatible with the license of this repository?

Good point. The icons are from the [Tango icon library](https://www.iconfinder.com/iconsets/tango-icon-library) licensed under [Attribution-Non-Commercial 3.0 Netherlands](https://creativecommons.org/licenses/by-nc/3.0/nl/deed.en_GB). So I guess we would need to either remove the icons or find MIT license icons, wouldn’t we?

> (I'd just fix these things, but I can't because it's a google drawing -- this is why I don't want it to be a google drawing!!!)

Wait, are you sure you are not able to access the Google Drawing? We have already updated the permissions. Can you try again? I just tried opening it in an incognito window and I am able to edit! The link is: https://docs.google.com/drawings/d/1FewA-gmh7vjuatHQlM_u7N56riyLfoH12huVOU0v1MA/edit

> I didn't realize hand-SVG was that difficult :S isn't it basically just html?

Hand-coding that diagram is very tiresome and time-consuming!

Reply:
I can edit it now!  But so can anybody on the internet -- that's probably worse :P

Reply:
Shall we just commit the image file as binary and say something like this image was generated from this Google Drawing?

Reply:
idk I still think making a simple svg wouldn't be that hard.  Sadly the one google spits out is kind of garbage.

It shouldn't be that bad though, it's a bunch of `<rect>`s and a few `<text>`s and some lines -- maybe I'll hack on svg-ifying it?

Reply:
(I've been reading through [this tutorial](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorial) btw for other reasons -- seems not so bad!)

Reply:
yeah turns out it's not so bad, I've started laying things out:

(ignore the blue background, I wanted to see whether it was transparent)

![](https://i.fluffy.cc/4sPwMtg532FtfS8gncfwNDGTvgpN4R3P.png)

Reply:
I appreciate your efforts in hand-coding the SVG but let’s take a step back and think about how often this diagram is going to be edited. I don’t expect too many changes happening in the diagram. Exporting a PNG and checking in the image is not a bad idea. It saves time and there is no need to duplicate the effort. Editing the Google Drawing later is also a much lower barrier of entry for others who would like to contribute.

Reply:
- the svg is very easy to edit
- I don't want to check in a binary
- google drive is a pretty high barrier to entry

Reply:
Okay. Please check-in the SVG in this branch once you’re done. I’d like to help with the styling.

Reply:
Here's my most recent version:

![](https://i.fluffy.cc/5b7TRr51s1g3TRlbfPSq2C9xvJ6F8mVs.svg)

link to source: https://i.fluffy.cc/5b7TRr51s1g3TRlbfPSq2C9xvJ6F8mVs.svg

feel free to tweak and commit it :)

(I tried to make the boxes as reusable as possible, but the lines are all ad-hoc sadly -- couldn't figure out quite how to make them click)

Reply:
That’s great! Sure, will tweak and commit.